### PR TITLE
[pat] Implement GetPersonalAccessToken

### PR DESCRIPTION
**Pull request by laushinka**
## Description
Implements GetPersonalAccessToken

## Related Issue(s)

Fixes #

## How to test
Unit tests `personal_access_token_test.go` and `tokens_test.go`

## Release Notes

```release-note
NONE
```

## Werft options:

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

**Comment by roboquat** [comment]
@laushinka: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

File: components/public-api-server/pkg/apiv1/tokens_test.go [hunk 1]
```diff
@@ -65,18 +65,39 @@ func TestTokensService_CreatePersonalAccessTokenWithoutFeatureFlag(t *testing.T)
...
+			Token: &v1.PersonalAccessToken{
+				Id:             response.Msg.Token.Id,
+				Hash:           response.Msg.Token.Hash,
+				Name:           response.Msg.Token.Name,
+				Description:    response.Msg.Token.Description,
+				ExpirationTime: response.Msg.Token.ExpirationTime,
+				Scopes:         response.Msg.Token.Scopes,
+				CreatedAt:      response.Msg.Token.CreatedAt,
+			},
+		}, response.Msg)
```
Comment (easyCZ) [inline components/public-api-server/pkg/apiv1/tokens_test.go:100]: After the rebase, you can now reuse the `personalAccessTokenToAPI` method. See https://github.com/gitpod-io/gitpod/pull/14832/files#diff-faefbff9f99df138bade476286958b2870dacd62989297aeef83b576e5954b8bR182

Comment (easyCZ) [inline components/public-api-server/pkg/apiv1/tokens_test.go:100]: This test shouldn't be under the test case for when the feature flag is disabled

File: components/public-api-server/pkg/apiv1/tokens.go [hunk 2]
```diff
@@ -65,13 +65,32 @@ func (s *TokensService) GetPersonalAccessToken(ctx context.Context, req *connect
...
+
+	return connect.NewResponse(&v1.GetPersonalAccessTokenResponse{Token: &v1.PersonalAccessToken{
+		Id:             response.ID.String(),
+		Hash:           response.Hash,
+		Name:           response.Name,
+		Description:    response.Description,
+		ExpirationTime: timestamppb.New(response.ExpirationTime),
+		Scopes:         response.Scopes,
+		CreatedAt:      timestamppb.New(response.CreatedAt),
+	}}), nil
```
Comment (easyCZ) [inline components/public-api-server/pkg/apiv1/tokens.go:93]:
```suggestion
	return connect.NewResponse(&v1.GetPersonalAccessTokenResponse{Token: personalAccessTokenToAPI(response)}), nil
```

File: components/public-api-server/pkg/apiv1/tokens.go [hunk 3]
```diff
@@ -65,13 +65,32 @@ func (s *TokensService) GetPersonalAccessToken(ctx context.Context, req *connect
...
 
 	log.Infof("Handling GetPersonalAccessToken request for Token ID '%s'", tokenID.String())
-	return nil, connect.NewError(connect.CodeUnimplemented, errors.New("gitpod.experimental.v1.TokensService.GetPersonalAccessToken is not implemented"))
+
+	userId, err := uuid.Parse(user.ID)
+	if err != nil {
+		return nil, connect.NewError(connect.CodeInternal, fmt.Errorf("Failed to parse user ID as UUID"))
+	}
+
+	response, err := db.GetToken(ctx, s.dbConn, tokenID, userId)
```
Comment (easyCZ) [inline components/public-api-server/pkg/apiv1/tokens.go:80]:
```suggestion
	token, err := db.GetToken(ctx, s.dbConn, tokenID, userId)
```

File: components/public-api-server/pkg/apiv1/tokens.go [hunk 4]
```diff
@@ -65,13 +65,32 @@ func (s *TokensService) GetPersonalAccessToken(ctx context.Context, req *connect
...
+	if err != nil {
+		connect.NewError(connect.CodeInternal, fmt.Errorf("Failed to get Personal Access Token for User %s", userId.String()))
+	}
```
Comment (easyCZ) [inline components/public-api-server/pkg/apiv1/tokens.go:83]: You also need to handle the NotFound case

File: components/gitpod-db/go/personal_access_token_test.go [hunk 5]
```diff
@@ -19,24 +19,33 @@ import (
...
+		_, err := db.GetToken(context.Background(), conn, token2.ID, token.UserID)
+		require.Error(t, err, "Token not found")
+	})
+
+	t.Run("valid", func(t *testing.T) {
+		returned, err := db.GetToken(context.Background(), conn, token.ID, token.UserID)
+		require.NoError(t, err)
+		require.Equal(t, token.ID, returned.ID)
+		require.Equal(t, token.UserID, returned.UserID)
+	})
```
Comment (easyCZ) [inline components/gitpod-db/go/personal_access_token_test.go:48]: Missing test case for when both the user, and tokenID do not exist in the DB

File: components/gitpod-db/go/personal_access_token_test.go [hunk 6]
```diff
@@ -19,24 +19,33 @@ import (
...
+	token := dbtest.NewPersonalAccessToken(t, db.PersonalAccessToken{UserID: firstUserId})
+	token2 := dbtest.NewPersonalAccessToken(t, db.PersonalAccessToken{UserID: secondUserId})
+
+	tokenEntries := []db.PersonalAccessToken{token, token2}
+
+	dbtest.CreatePersonalAccessTokenRecords(t, conn, tokenEntries...)
+
+	t.Run("not matching user", func(t *testing.T) {
+		_, err := db.GetToken(context.Background(), conn, token.ID, token2.UserID)
+		require.Error(t, err, "Token not found")
```
Comment (easyCZ) [inline components/gitpod-db/go/personal_access_token_test.go:34]: GetToken should return a `db.NotFoundError`, which you should match on in the error. See other usage of NotFoundError in the db package

File: components/gitpod-db/go/personal_access_token.go [hunk 7]
```diff
@@ -36,14 +35,14 @@ func (d *PersonalAccessToken) TableName() string {
 	return "d_b_personal_access_token"
 }
 
-func GetToken(ctx context.Context, conn *gorm.DB, id uuid.UUID) (PersonalAccessToken, error) {
+func GetToken(ctx context.Context, conn *gorm.DB, id uuid.UUID, userId uuid.UUID) (PersonalAccessToken, error) {
```
Comment (easyCZ) [inline components/gitpod-db/go/personal_access_token.go:38]:
```suggestion
func GetTokenForUser(ctx context.Context, conn *gorm.DB, tokenID uuid.UUID, userID uuid.UUID) (PersonalAccessToken, error) {
```

File: components/gitpod-db/go/personal_access_token.go [hunk 8]
```diff
@@ -36,14 +35,14 @@ func (d *PersonalAccessToken) TableName() string {
...
 	var token PersonalAccessToken
 
 	db := conn.WithContext(ctx)
 
-	db = db.Where("id = ?", id).First(&token)
+	db = db.Where("id = ?", id).Where("userId", userId).First(&token)
 	if db.Error != nil {
-		return PersonalAccessToken{}, fmt.Errorf("Failed to retrieve token: %w", db.Error)
+		return PersonalAccessToken{}, fmt.Errorf("Failed to retrieve token: %v", db.Error)
 	}
```
Comment (easyCZ) [inline components/gitpod-db/go/personal_access_token.go:46]: Explicitly handle NotFound, and return db.NotFoundError

**Review by easyCZ — APPROVED** [review]

**Comment by laushinka** [comment]
Hold to fix failing build.

/hold

**Pull request merged**